etl/steps/data/meadow/iea/2024-07-04/critical_minerals.py

```diff
@@ -0,0 +1,245 @@
+    tb = tb[~mineral_mask].reset_index(drop=True)
+
+    return tb
+
+
+def process_demand_for_key_minerals(
+    data: pr.ExcelFile, sheet_name: str, header_row_name: Optional[str], normal_row_name: str, tb_short_name: str
+) -> Table:
+    tb = data.parse(sheet_name)
+
+    scenarios = {
+        "": "",
+        "Stated Policies scenario": "Stated policies",
+        "Announced Pledges scenario": "Announced pledges",
+        "Net Zero Emissions by 2050 scenario": "Net zero by 2050",
+    }
+    scenarios_columns = tb.iloc[2].ffill().fillna("").values
+    assert set(scenarios) == set(scenarios_columns), f"Format has changed for sheet: {sheet_name}"
+
+    # Define new columns.
+    new_columns = (
+        tb.iloc[2].ffill().fillna("").replace(scenarios).values
+        + "__"
+        + tb.iloc[3].astype("Int64").astype(object).fillna("").astype(str).values
+    )
+    new_columns[0] = normal_row_name
+    new_columns[1] = f"Current{new_columns[1]}"
+    tb = tb.rename(columns={tb.columns[i]: column for i, column in enumerate(new_columns)}, errors="raise")
+
+    # By construction, all columns that end in "__" do not correspond to any data, so drop them.
+    tb = tb.drop(columns=[column for column in tb.columns if column.endswith("__")])
+
+    # Drop initial rows that corresponded to titles.
+    tb = tb.drop([0, 1, 2, 3, 4]).reset_index(drop=True)
+
+    # Drop empty columns and rows.
+    tb = tb.dropna(axis=1, how="all").dropna(how="all").reset_index(drop=True)
+
+    # Drop rows containing notes at the end of the sheet.
+    tb = drop_all_rows_of_notes(tb=tb)
+
+    if header_row_name is not None:
+        # Separate minerals and technology in different columns.
+        tb = transform_header_rows_into_new_column(
+            tb=tb, header_row_name=header_row_name, normal_row_name=normal_row_name
+        )
+        main_index_columns = [header_row_name, normal_row_name]
+    else:
+        main_index_columns = [normal_row_name]
+
+    # Transform table to end up with another table that has a "year" and "scenario" columns.
+    scenarios_names = ["Current"] + [scenario for scenario in scenarios.values() if scenario != ""]
+    tables = []
+    for scenario in scenarios_names:
+        _tb = tb[main_index_columns + [column for column in tb.columns if column.startswith(scenario)]]
+        _tb = _tb.rename(columns={column: column.replace(f"{scenario}__", "") for column in _tb.columns})
+        _tb = _tb.melt(id_vars=main_index_columns, value_name="demand", var_name="year")
+        _tb = _tb.assign(**{"scenario": scenario})
+        tables.append(_tb)
+    tb = pr.concat(tables, short_name=tb_short_name)
+
+    # Ensure all columns are snake-case, set an appropriate index, and sort conveniently.
+    tb = tb.format(main_index_columns + ["year", "scenario"])
+
+    return tb
+
+
+def process_supply_for_key_minerals(data: pr.ExcelFile) -> Table:
+    sheet_name = "2 Total supply for key minerals"
```

Comment:
If the sheet name changes, the error is very easy to spot, it's "ValueError: Worksheet named '...' not found", so including an assertion doesn't add much. If the structure of the sheet changes in any other way (like not having a separating column) it's more likely that the code will fail in various ways rather than it will generate a faulty table, so I think this function is "robust enough".
I've added a docstring to this file!